### PR TITLE
[API][Order] Filtering on the Orders list

### PR DESCRIPTION
**Pull request by LGouttefange**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13            |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | partially #11250                      |
| License         | MIT                                                          |

Included in this PR:
> Filtering on the Orders list
> All the features from features/order/managing_orders/order_filtration
> 12 scenarios, 3 files

As this is my first PR here, I am open to any feedback :pray:

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+final class InclusiveDateRangeFilter implements ContextAwareFilterInterface
+{
+    /**
+     * @param ContextAwareFilterInterface $dateFilter
+     * @param string[] $properties
+     */
+    public function __construct(private ContextAwareFilterInterface $dateFilter, private array $properties)
+    {
+    }
+
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php:34]: Please remove second empty line

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * To include a '1970-01-01 13:00', it transforms:
+ *      'before' => '1970-01-01'
+ * into:
+ *      'before' => '1970-01-01 23:59'
+ */
+final class InclusiveDateRangeFilter implements ContextAwareFilterInterface
+{
+    /**
+     * @param ContextAwareFilterInterface $dateFilter
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]: This `@param` is unnecessary, docblocs should only be added to cover arguable cases for static analysis.

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     * @param class-string $resourceClass
+     * @return array<string, mixed>
+     */
+    public function getDescription(string $resourceClass): array
+    {
+        return $this->dateFilter->getDescription($resourceClass);
+    }
+
+    /**
+     * @param QueryBuilder $queryBuilder
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]: I guess only
```
* @param class-string $resourceClass
* @param array<string, mixed> $context
```

are needed, And
```patch
--- * @param string $resourceClass
+++ * @param class-string $resourceClass
```

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    protected function makeDateInclusive(string $value): string
+    {
+        return (new \DateTimeImmutable($value))
+            ->setTime(23, 59)
+            ->format('Y-m-d H:i');
+    }
+
+    private function isExclusive(mixed $value): bool
+    {
+        if(!$value || !is_string($value)) {
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]:
```suggestion
        if (!$value || !is_string($value)) {
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 5]
```diff
@@ -24,6 +24,13 @@
                 <attribute name="path">/admin/orders</attribute>
                 <attribute name="filters">
                     <attribute>sylius.api.order_number_filter</attribute>
+                    <attribute>sylius.api.order_checkout_completed_at_filter</attribute>
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml]: Please sort them logically or alphabetically.

For example.
```xml
                    <attribute>sylius.api.order_number_filter</attribute>
                    <attribute>sylius.api.order_checkout_completed_at_filter</attribute>
                    <attribute>sylius.api.order_currency_code_filter</attribute>
                    <attribute>sylius.api.order_total_filter</attribute>
                    <attribute>sylius.api.order_channel_filter</attribute>
                    <attribute>sylius.api.order_shipping_method_filter</attribute>
                    <attribute>sylius.api.order_products_filter</attribute>
                    <attribute>sylius.api.order_variants_filter</attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 6]
```diff
@@ -218,6 +218,63 @@
...
+        </service>
+
+        <service id="sylius.api.order_channel_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
+            <argument type="collection">
+                <argument key="channel.code">exact</argument>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_products_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_products_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 7]
```diff
@@ -218,6 +218,63 @@
...
+
+        <service id="sylius.api.order_checkout_completed_at_filter" class="Sylius\Bundle\ApiBundle\Filter\InclusiveDateRangeFilter" public="true">
+            <argument type="service" id="sylius.api.order_checkout_completed_at_filter.inner" />
+            <argument type="collection">
+                <argument>checkoutCompletedAt</argument>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_channel_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_channel_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 8]
```diff
@@ -218,6 +218,63 @@
...
+        </service>
+
+        <service id="sylius.api.order_products_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
+            <argument type="collection">
+                <argument key="items.variant.product">exact</argument>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_shipping_method_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_shipping_method_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 9]
```diff
@@ -218,6 +218,63 @@
...
+            <argument type="collection">
+                <argument key="shipments.method.code">exact</argument>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_currency_code_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_currency_code_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 10]
```diff
@@ -218,6 +218,63 @@
...
+            <argument type="collection">
+                <argument key="currencyCode">exact</argument>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_total_filter" parent="api_platform.doctrine.orm.range_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_total_filter" parent="api_platform.doctrine.orm.range_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 11]
```diff
@@ -218,6 +218,63 @@
...
+            <argument type="collection">
+                <argument key="total"/>
+            </argument>
+            <tag name="api_platform.filter" />
+        </service>
+
+        <service id="sylius.api.order_variants_filter" parent="api_platform.doctrine.orm.search_filter"  public="true">
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.api.order_variants_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+     * @param array<string, mixed> $context
+     * @return void
+     *
+     * @throws \Exception
+     */
+    public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null, array $context = []): void
+    {
+        foreach ($this->properties as $property) {
+            $dateBefore = $context['filters'][$property]['before'] ?? null;
+            if ($this->isExclusive($dateBefore)) {
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]:
```suggestion
            if ($this->hasTime($dateBefore)) {
```

Function checks for time presence, so it should be named accordingly.

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        return (new \DateTimeImmutable($value))
+            ->setTime(23, 59)
+            ->format('Y-m-d H:i');
+    }
+
+    private function isExclusive(mixed $value): bool
+    {
+        if(!$value || !is_string($value)) {
+            return false;
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]: Function named `isSomething()` shouldn't return false on invalid argument and throw exception instead.

I guess it's not a option for this case, so how about `$this->getDateTime(?string $date, string $defaultTime): ?string`

https://github.com/Sylius/Grid/blob/ca081e180bf1eaa25a7fc4e9493bd5b88a95a0cc/Filter/DateFilter.php#L67-L77

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * When given a date-only "before" value to a DateFilter, allow it to include dates on that day
+ *
+ * To include a '1970-01-01 13:00', it transforms:
+ *      'before' => '1970-01-01'
+ * into:
+ *      'before' => '1970-01-01 23:59'
+ */
+final class InclusiveDateRangeFilter implements ContextAwareFilterInterface
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php:26]: `Inclusive` has two meaning: before (lte) is already inclusive to strictly_before (lt) as well as time inclusive `23:59`

So I'm not sure about naming, but don't have better idea.

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 15]
```diff
@@ -218,6 +218,63 @@
             <tag name="api_platform.filter" />
         </service>
 
+
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml]: Please remove unnecessary second empty line.

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 16]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
 
         Assert::same($firstItem['number'], str_replace('#', '', $number));
     }
+
+
+    /**
+     * @When I specify filter date from as :dateTime
+     */
+    public function iSpecifyFilterDateFromAs($dateTime)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: Please use new `Context/Api` as example, it's covered with type hints.

```suggestion
    public function iSpecifyFilterDateFromAs(string $dateTime): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 17]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->addFilter('checkoutCompletedAt[after]', $dateTime);
+    }
+
+    /**
+     * @When I specify filter date to as :dateTime
+     */
+    public function iSpecifyFilterDateToAs($dateTime)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iSpecifyFilterDateToAs(string $dateTime): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 18]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->addFilter('checkoutCompletedAt[before]', $dateTime);
+    }
+
+    /**
+     * @When I filter
+     */
+    public function iFilter()
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iFilter(): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 19]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->filter();
+    }
+
+    /**
+     * @Then I should see :count orders in the list
+     */
+    public function itShouldHaveAmountOfItemsInTheList($count)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function itShouldHaveAmountOfItemsInTheList(int $count): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 20]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $numberOfItems = $this->responseChecker->getValue($this->client->getLastResponse(), 'hydra:totalItems');
+        Assert::eq($numberOfItems, $count);
+    }
+
+    /**
+     * @Then I should not see an order with :number number
+     */
+    public function iShouldNotSeeAnOrderWithNumber($number)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iShouldNotSeeAnOrderWithNumber(int $number): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 21]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        Assert::false($this->responseChecker->hasItemWithValue($this->client->getLastResponse(), 'number', $number));
+    }
+
+    /**
+     * @When I choose :channel as a channel filter
+     */
+    public function iChooseChannelAsAChannelFilter(ChannelInterface $channel)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iChooseChannelAsAChannelFilter(ChannelInterface $channel): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 22]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    public function iFilterByProduct(ProductInterface $product): void
+    {
+        $this->client->addFilter('items.variant.product[]', $product->getId());
+
+        $this->iFilter();
+    }
+
+    /**
+     * @When /^I filter by (products "([^"]+)" and "([^"]+)")$/
+     * @param ProductInterface[] $products
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: Please remove this line, unless necessary.

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 23]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+            $this->client->addFilter('items.variant.product[]', $productName->getId());
+        }
+
+        $this->iFilter();
+    }
+
+    /**
+     * @When I choose :shippingMethod as a shipping method filter
+     */
+    public function iChooseMethodAsAShippingMethodFilter(ShippingMethodInterface $shippingMethod)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iChooseMethodAsAShippingMethodFilter(ShippingMethodInterface $shippingMethod): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 24]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->addFilter('shipments.method.code', $shippingMethod->getCode());
+    }
+
+
+    /**
+     * @When I choose :currency as the filter currency
+     */
+    public function iChooseAsTheFilterCurrency(CurrencyInterface $currency)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iChooseAsTheFilterCurrency(CurrencyInterface $currency): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 25]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->addFilter('currencyCode', $currency->getCode());
+    }
+
+    /**
+     * @Then I should not see any orders with currency :currency
+     */
+    public function iShouldNotSeeAnyOrdersWithCurrency(CurrencyInterface $currency)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iShouldNotSeeAnyOrdersWithCurrency(CurrencyInterface $currency): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 26]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        Assert::false($this->responseChecker->hasItemWithValue($this->client->getLastResponse(), 'currencyCode', $currency->getCode()));
+    }
+
+    /**
+     * @When I specify filter total being greater than :total
+     */
+    public function iSpecifyFilterTotalBeingGreaterThan($total)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: float/string $total?

```suggestion
    public function iSpecifyFilterTotalBeingGreaterThan($total): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 27]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    {
+        $this->client->addFilter('total[gt]', $total* 100);
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
        $this->client->addFilter('total[gt]', $total * 100);
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 28]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+    }
+
+    /**
+     * @When I specify filter total being less than :total
+     */
+    public function iSpecifyFilterTotalBeingLessThan($total)
+    {
+        $this->client->addFilter('total[lt]', $total* 100);
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
        $this->client->addFilter('total[lt]', $total * 100);
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 29]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+     */
+    public function iSpecifyFilterTotalBeingGreaterThan($total)
+    {
+        $this->client->addFilter('total[gt]', $total* 100);
+    }
+
+    /**
+     * @When I specify filter total being less than :total
+     */
+    public function iSpecifyFilterTotalBeingLessThan($total)
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
    public function iSpecifyFilterTotalBeingLessThan(float/string? $total): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 30]
```diff
@@ -345,4 +350,140 @@ public function theFirstOrderShouldHaveNumber(string $number)
...
+     */
+    public function iFilterByVariant(ProductVariantInterface $variant): void
+    {
+        $this->client->addFilter("items.variant[]", $variant->getId());
+        $this->client->filter();
+    }
+
+    /**
+     * @When /^I filter by (variants "([^"]+)" and "([^"]+)")$/
+     * @param ProductVariantInterface $variants
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: Please remove this line unless necessary by phpstan

**Review by diimpp — CHANGES_REQUESTED** [review]

**Comment by diimpp** [comment]
@LGouttefange Hi, great filtering coverage. :+1: I've commented on a few minor points, please take a look.

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php (hunk 14 above)
Comment (LGouttefange) [reply src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php:26]: I agree. In this edge case of adapting the DateFilter to solve the `Filtering orders by date from without time` test, naming it is quite tricky haha.

Perhaps naming it `DateWithoutTimeFilter` would make it less confusing ? As well as adding a point in the OpenAPI documentation that it behaves that way, and not like a standard ApiPlatform DateFilter.

File: src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php (hunk 13 above)
Comment (LGouttefange) [reply src/Sylius/Bundle/ApiBundle/Filter/InclusiveDateRangeFilter.php]: I've pushed a new version of the Filter class, taking inspiration of the mentioned filter.
In the case of a Filter decorator, I didn't want to overstep or reproduce the base DateFilter validation & error handling, hence why I am not actively validating the item and just returning it as such.

**Review by LGouttefange — COMMENTED** [review]
Thank you @diimpp for the thorough review. I've updated & commented some points
I also fixed the coding standards issues, but -  bit of a newbie question - please let me know if I should resolve them myself, or let you handle those :pray: thanks !

**Comment by jakubtobiasz** [comment]
Closing as already done in #15566. Thanks!
